### DISCUSSION TITLE
How to merge objects in a collection with the same key

### DISCUSSION THREAD

**Discussion by jwanggg** [open-policy-agent/community#281] [OPA and Rego]
Hi I was looking at the object comprehension in order to decompose objects and transform them but am not getting a correct result.

Given my data

```
[{
		"input1": [
			"a",
			"b"
		]
	},
	{
		"input1": [
			"a",
			"c",
			"d"
		]
	},
	{
		"input2": [
			"a"
		]
	}
]

```

How would I transform it into just

```
[{
		"input1": [
			"a",
			"b",
			"c",
			"d"
		]
	},
	{
		"input2": [
			"a"
		]
	}
]
```

I tried with a more complex dataset and query here https://play.openpolicyagent.org/p/Pw6kSPNU4R. Not sure why it doesn't merge together when the keys are the same as well

**Comment by johanfylling** [comment]
Hi Jonathan!

I'm a bit rusty, so there are likely more elegant solutions, but this should do the trick:

```rego
package play

merged[x] {                            # 4. putting all objects in a set
    input[_][key]                      # 1. enumerating the unique keys
    set := {v | v := input[_][key][_]} # 2. joining all list items with the same key into a set
    x := {key: set}                    # 3. making an object out of the key and associated set
}
```
https://play.openpolicyagent.org/p/H5fHwLc6yR

**Reply by jwanggg** [reply]
Thanks @johanfylling I tried with a more complex data sample and input here https://play.openpolicyagent.org/p/Pw6kSPNU4R. Seems like they don't merge correctly. I wonder if it's because it is too nested? Trying to merge the first application keys and then merge the experiment part as well so it's like twice

**Reply by johanfylling** [reply]
I don't quite follow what you're trying to accomplish. Could you please provide a sample of your expected output?

**Reply by jwanggg** [reply]
Sorry for the late reply, essentially the above example I wanted the two experiment keys merged. Here's an updated policy where the permissions in the output is what I wanted https://play.openpolicyagent.org/p/jMmZmU6zTX. Permissions0 and Permissions1 are the steps before merging into getting it into the final Permissions. Wondering if there was a way to simplify everything.

**Comment by anderseknert** [comment]
For simple merges you have the [object.union](https://www.openpolicyagent.org/docs/latest/policy-reference/#builtin-object-objectunion) and [object.union_n](https://www.openpolicyagent.org/docs/latest/policy-reference/#builtin-object-objectunion_n) built-in functions to work with. There was a recent proposal by @charlesdaniels to add another one with [object.union_merge](https://github.com/open-policy-agent/opa/issues/5144), but as your requirements show, there aren't really any agreed-upon resolutions for merging values like arrays... so we'd either need to have one built-in per option, or provide some generic way of bringing in your own "conflict resolver".. which would be a challenge in a non-functional language like Rego, I think.

So for the time being, you're probably best off with a custom implemnentation like the one @johanfylling provided here.
